Providers/SyncStatusResolver.swift

```diff
@@ -0,0 +1,110 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+import Foundation
+import Sync
+import XCGLogger
+import Deferred
+import Shared
+import Storage
+
+public enum SyncDisplayState {
+    case InProgress
+    case Good
+    case Bad(message: String?)
+    case Stale(message: String)
+
+    func asObject() -> [String: String]? {
+        switch self {
+        case .Bad(let msg):
+            guard let message = msg else {
+                return ["state": "Error"]
+            }
+            return ["state": "Error",
+                    "message": message]
+        case .Stale(let message):
+            return ["state": "Warning",
+                    "message": message]
+        default:
+            break
+        }
+        return nil
+    }
+}
+
+public func ==(a: SyncDisplayState, b: SyncDisplayState) -> Bool {
+    switch (a, b) {
+    case (.InProgress,   .InProgress): return true
+    case (.Good,   .Good): return true
+    case (.Bad(let a), .Bad(let b)) where a == b: return true
+    case (.Stale(let a), .Stale(let b)) where a == b: return true
+    default: return false
+    }
+}
+
+private let log = Logger.syncLogger
+
+public struct SyncStatusResolver {
+    let engineResults: Maybe<EngineResults>
+
+    public func resolveResults() -> SyncDisplayState {
+        guard let results = engineResults.successValue else {
+            switch engineResults.failureValue {
+            case _ as BookmarksMergeError, _ as BookmarksDatabaseError:
+                return SyncDisplayState.Stale(message: String(format: Strings.FirefoxSyncPartialTitle, Strings.localizedStringForSyncComponent("bookmarks") ?? ""))
+            default:
+                return SyncDisplayState.Bad(message: nil)
+            }
+        }
+
+        // Run through the engine results and produce a relevant display status for each one
+        let displayStates: [SyncDisplayState] = results.map { (engineIdentifier, syncStatus) in
+            log.debug("Sync status for \(engineIdentifier): \(syncStatus)")
+
+            switch syncStatus {
+            case .NotStarted(let reason):
+                switch reason {
+
+                // Potentially problematic reasons for not starting. Should surface to the user.
```

Comment:
I feel like some of these depend on the application state. If I tap "Sync Now" and anything bad happens, tell me. An automatic sync shouldn't complain because I'm offline!